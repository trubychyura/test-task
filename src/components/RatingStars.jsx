import { useState } from 'react';
import { Form } from 'react-bootstrap';
import { Field } from 'formik';

import {
  StyledContainer,
  StyledRadio,
  StyledStar,
} from '../styled/RatingStars';

const RatingStars = ({ stars, name }) => {
  const [hover, setHover] = useState(null);

  if (stars !== undefined) {
    return (
      <StyledContainer>
        {[...Array(5)].map((_, i) => (
          <StyledStar
            value={i + 1}
            rating={Math.round(stars)}
            hover={null}
            key={i}
          />
        ))}
      </StyledContainer>
    );
  }

  return (
    <StyledContainer>
      <Field name={name}>
        {({ field }) =>
          [...Array(5)].map((_, i) => {
            let rating = i + 1;
            return (
              <Form.Label htmlFor={`formRadio${i}`} key={i}>
                <StyledStar
                  value={rating}
                  rating={field.value}
                  hover={hover}
                  onMouseEnter={() => setHover(rating)}
                  onMouseLeave={() => setHover(0)}
                />
                <StyledRadio
                  type='radio'
                  id={`formRadio${i}`}
                  {...field}
                  value={rating}
                  checked={rating === field.value}
                />
              </Form.Label>
            );
          })
        }
      </Field>
    </StyledContainer>
  );
};

export default RatingStars;
