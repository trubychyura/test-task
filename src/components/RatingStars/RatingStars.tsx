import { FC, useState } from 'react';
import { Field, FieldProps } from 'formik';
import { Form } from 'react-bootstrap';

import { RatingProps } from '../../types';
import {
  StyledContainer,
  StyledRadio,
  StyledStar,
} from '../../styled/StyledRating';

const RatingStars: FC<RatingProps> = ({ name, stars }) => {
  const [hover, setHover] = useState<number | null>(null);

  if (stars) {
    return (
      <StyledContainer>
        {[...Array(5)].map((_, i) => (
          <StyledStar value={i + 1} rating={Math.round(stars || 0)} key={i} />
        ))}
      </StyledContainer>
    );
  }

  return (
    <StyledContainer>
      <Field name={name}>
        {({ field }: FieldProps) =>
          [...Array(5)].map((_, i) => {
            let rating: number = i + 1;

            return (
              <Form.Label htmlFor={`formRadio${i}`} key={i}>
                <StyledStar
                  value={rating}
                  rating={field.value}
                  hover={hover}
                  onMouseEnter={() => setHover(rating)}
                  onMouseLeave={() => setHover(0)}
                />
                <StyledRadio
                  type='radio'
                  id={`formRadio${i}`}
                  {...field}
                  value={rating}
                  checked={rating === field.value}
                  data-testid={`radio${i}`}
                />
              </Form.Label>
            );
          })
        }
      </Field>
    </StyledContainer>
  );
};

export default RatingStars;
