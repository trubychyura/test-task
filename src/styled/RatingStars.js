import styled from 'styled-components';
import { Star } from '@styled-icons/boxicons-solid/Star';
import { Form } from 'react-bootstrap';

export const StyledStar = styled(Star)`
  cursor: pointer;
  transition: color 200ms;

  ${({ value, rating, hover }) =>
    value <= (hover || rating) ? `color:  gold;` : `color: silver;`}
`;

export const StyledRadio = styled(Form.Check)`
  display: none;
`;

export const StyledContainer = styled(Form.Group)`
  display: flex;
`;
