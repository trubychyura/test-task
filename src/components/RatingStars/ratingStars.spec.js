import { shallow } from 'enzyme';

import RatingStars from '..';
import { Field } from 'formik';

const setUp = (props) => shallow(<RatingStars {...props} />);

describe('RatingStars component', () => {
  let component;
  beforeEach(() => {
    component = setUp();
  });

  it('should contain a Field', () => {
    expect(component.find(Field)).toHaveLength(1);
  });

  it('should contain 5 children', () => {
    const component = setUp({ stars: 5 });
    expect(component.children()).toHaveLength(5);
  });

  it('should render component with 5 children', () => {
    const component = setUp({ stars: 5 });
    expect(component).toMatchSnapshot();
  });

  it('should render component with Field', () => {
    expect(component).toMatchSnapshot();
  });
});
