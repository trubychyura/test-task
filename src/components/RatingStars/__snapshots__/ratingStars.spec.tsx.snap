// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingStars component should render component with 5 children 1`] = `
<styled.div>
  <Styled(Star)
    key="0"
    rating={5}
    value={1}
  />
  <Styled(Star)
    key="1"
    rating={5}
    value={2}
  />
  <Styled(Star)
    key="2"
    rating={5}
    value={3}
  />
  <Styled(Star)
    key="3"
    rating={5}
    value={4}
  />
  <Styled(Star)
    key="4"
    rating={5}
    value={5}
  />
</styled.div>
`;

exports[`RatingStars component should render component with Field 1`] = `
<styled.div>
  <Field>
    <Component />
  </Field>
</styled.div>
`;
